### DISCUSSION TITLE
Reassigning Ctrl+x ?

### DISCUSSION THREAD

**Discussion by fiso64** [chrisant996/clink#662] [Q&A] [answered]
I'm wondering if it's possible to reassign Ctrl-x to another action (like `cua-cut`). I've tried adding a line with "\C-X" to my inputrc, but it doesn't seem to work because it waits for the next key press as there are some default shortcuts like Ctrl-x,Ctrl-u.

**Comment by chrisant996** [comment]
You will need to undefine all of the <kbd>Ctrl-X</kbd> chords, i.e. all of the multi-key bindings that start with <kbd>Ctrl-X</kbd>.

**Reply by fiso64** [reply]
How can I undefine them?

**Reply by chrisant996** [reply]
### How can I unbind a key?

Undefine a key by binding it to nothing.

Example of defining a key binding:
```inputrc
"\C-x\C-r":    clink-reload
```

Example of undefining the same key binding:
```inputrc
"\C-x\C-r":
```

### How can I find what keys are bound?

There are many built-in ways to find what keys are bound, but most of them are designed around using keys, rather than raw configuring of .inputrc files.  None of the ways can be redirected to a file, but you could copy/paste from the terminal output.

Pressing <kbd>Alt-1</kbd><kbd>Ctrl-X</kbd><kbd>Ctrl-F</kbd> lists the current key bindings in the same format as used by `.inputrc` files.  But the list is sorted by command name, not by key name.  To find keys starting with <kbd>Ctrl-X</kbd> you'd need to filter the list, e.g. piping it through `findstr "C-x,"`.

Or here is a Lua script that creates a new command and binds it to <kbd>Alt-Ctrl-X</kbd>.  The command asks you to press a key, and then lists all key bindings that start with that key.  You could load that script, press <kbd>Alt-Ctrl-X</kbd> and then <kbd>Ctrl-X</kbd> and then copy/paste the leftmost column of text from the terminal, and that would be a list of "unbind" key bindings for the keys.

The Readline library used by Clink doesn't make it easy or convenient to do sophisticated configuration of key bindings, and that's a side effect of its Unix terminal roots.

```lua
rl.setbinding([["\M-\C-x"]], [["luafunc:list_bindings_for_key"]])

function list_bindings_for_key(rl_buffer)
    rl_buffer:beginoutput()

    clink.print("Press a key:  ", NONL)
    local key = console.readinput(false)
    if not key then
        print()
        print("Unable to read key.")
        return
    end

    clink.print("\x1b[1m"..rl.translatekey(key, 1).."\x1b[m")

    local binding = rl.translatekey(key, 2)
    if not binding then
        print("Unrecognized key.")
        return
    end

    local stripped = binding:match('^(".*)"$')
    binding = stripped or binding

    for _,kb in ipairs(rl.getkeybindings(true)) do
        if kb.key:sub(1, #binding) == binding then
            local s = kb.key:gsub(" +$", "")..":"
            s = string.format("%-22s  %s", s, kb.binding)
            print(string.format("%-62s  # %s", s, rl.translatekey(kb.key, 3)))
        end
    end
end
```

**Reply by fiso64** [reply]
I used the script to print a list of bindings starting with Ctrl-x, and added the following to my inputrc:
```
"\C-x(":
"\C-x)":
"\C-x*":
"\C-x\C-?":
"\C-x\C-e":
"\C-x\C-g":
"\C-x\C-r":
"\C-x\C-u":
"\C-x\C-x":
"\C-x\C-z":
"\C-xe":
"\C-xg":
"\C-x"				cua-cut
```
Unfortunately this still does not work. Running the script again only lists the single cua-cut binding for Ctrl-x.

**Reply by chrisant996** [reply]
You forgot the `:`.  It's not optional.

The "cua-cut" line is invalid because it doesn't follow the syntax.; it's missing the `:` that indicates the end of the key sequence, and the beginning of the command to bind to the key sequence.

The syntax is documented [here](https://chrisant996.github.io/clink/clink.html#readline-key-bindings) (Clink docs), and also [here](https://www.gnu.org/software/bash/manual/html_node/Readline-Init-File-Syntax.html) (bash docs) and [here](https://tiswww.case.edu/php/chet/readline/readline.html#Readline-Init-File-Syntax) (Readline docs).

(I didn't invent this syntax; it's existed this way in bash and Readline for a few decades.)

**Reply by fiso64** [reply]
I did forget it, my bad. interestingly, bindings still seem to work even without the `:`, I just tested it with a few other bindings.

Anyways, after adding the `:`, it's still not working for me.

**Reply by chrisant996** [reply]
Ah ... right, I just tested this.

This is a known limitation in Readline (and bash and all programs that use Readline).

Once a multi-key sequence has been defined, there is no way to undefine the multi-ness and go back to a single key.

Readline itself defines <kbd>Ctrl-X</kbd><kbd>Ctrl-F</kbd> and other combinations starting with <kbd>Ctrl-X</kbd>.

This was discussed in issue in the Clink repo in the past, and I'm looking for the issue number to link it here.
_**UPDATE:**  I found it.  This unbind issue was discussed in https://github.com/chrisant996/clink/issues/248#issuecomment-1055050993, in the context of a different kind of situation.  See the final section of that comment ("Nope, there is a Readline problem").  But for that to make sense, also read the earlier section of that comment, "Sidebar: What is a shadowed key?"._

**Reply by chrisant996** [reply]
I can't find the issue with the previous discussion.
_**UPDATE:**  I found it.  This unbind issue was discussed in https://github.com/chrisant996/clink/issues/248#issuecomment-1055050993, in the context of a different kind of situation.  See the final section of that comment ("Nope, there is a Readline problem").  But for that to make sense, also read the earlier section of that comment, "Sidebar: What is a shadowed key?"._

But what happens is that <kbd>Ctrl-X</kbd> will do the `cua-cut` command ... but not until you press another key -- because Readline is stuck on believing that <kbd>Ctrl-X</kbd> is the beginning of a multi-key binding, since it doesn't know how to undo that once it's been done.

<kbd>Ctrl-X</kbd> is the only key that suffers that from that by default.  I can't remove all the <kbd>Ctrl-X</kbd> key bindings by default, as that would create an even worse problem.

I'll explore whether it might be reasonable to try to patch how Readline handles that situation, though.

**Reply by fiso64** [reply]
I see, thanks for answering

**Reply by chrisant996** [reply, marked as answer]
@fiso64 I've debugged through it, and now I see what's going on...

Readline has code that tries to revert the shadow key issue when unbinding keys.

But there are four caveats:
1. The code can only revert the shadow key issue when keys are bound to commands; not when they're bound to macros.  That doesn't appear to be relevant in your case.
2. The code can only revert one level of the shadow key issue at a time.  That isn't relevant in your case, because the longest <kbd>Ctrl-X</kbd> key binding is only 2 keys long.  But if there were any that were 3 or more long, then it would be a lot more complicated and persnickety about how to successfully fully revert them.
3. The code can only revert the shadow key issue when all sub-keys become unbound.  And while it looks to us like all the sub-keys have been unbound ... they actually haven't.  More on that in a moment.
4. Readline cheats and has a hard-coded table listing its default <kbd>Ctrl-X</kbd> key bindings.  So if you successfully unbind all of the <kbd>Ctrl-X</kbd> bindings, then Readline should try to free hard-coded non-allocated memory and it should either crash or corrupt memory.  _But Readline has a bug where it leaks the memory instead of freeing it, when unbinding the last key from a sub-table.  So two wrongs cancel each other out and it doesn't crash -- but it does leak a small amount of memory._

Ok, so caveats 1, 2, and 4 end up working out in your case.

But not caveat 3.  Because Readline has a hard-coded table of key bindings for <kbd>Ctrl-X</kbd>, and all of the uppercase letters are bound to the command [do-lowercase-version](https://chrisant996.github.io/clink/clink.html#rlcmd-do-lowercase-version) (i.e. <kbd>Ctrl-X</kbd><kbd>A</kbd> .. <kbd>Ctrl-X</kbd><kbd>Z</kbd>).  And that command doesn't get listed as being a key binding when retrieving key bindings, because the whole point is for it to silently do whatever is bound to the same _lowercase_ letter instead.

So to successfully fully revert the <kbd>Ctrl-X</kbd> key bindings, you have to also list all the upper case letter combinations.

**I confirmed that once they're all unbound as well, then it finally does what you wanted.**

What an ordeal, eh?!

```inputrc
# Unbind the keys that were explicitly bound.
"\C-x(":
"\C-x)":
"\C-x*":
"\C-x\C-?":
"\C-x\C-e":
"\C-x\C-g":
"\C-x\C-r":
"\C-x\C-u":
"\C-x\C-x":
"\C-x\C-z":
"\C-xe":
"\C-xg":

# Unbind the keys that were implicitly bound.
"\C-xA":
"\C-xB":
"\C-xC":
"\C-xD":
"\C-xE":
"\C-xF":
"\C-xG":
"\C-xH":
"\C-xI":
"\C-xJ":
"\C-xK":
"\C-xL":
"\C-xM":
"\C-xN":
"\C-xO":
"\C-xP":
"\C-xQ":
"\C-xR":
"\C-xS":
"\C-xT":
"\C-xU":
"\C-xV":
"\C-xW":
"\C-xX":
"\C-xY":
"\C-xZ":

# Finally, bind Ctrl-X.
"\C-x":			cua-cut
```

**Reply by fiso64** [reply]
Yep, it works! Thanks a lot for the effort

**Comment by chrisant996** [comment]
Note:  what you're trying to do is some relatively advanced configuration.  This is not a common thing.
